backend/tracim_backend/lib/utils/authentication.py

```diff
@@ -20,16 +28,195 @@
 from tracim_backend.models.auth import AuthType
 from tracim_backend.models.auth import User
 
+from saml2.config import Config as Saml2Config
+from saml2.metadata import create_metadata_string
+from saml2 import BINDING_HTTP_POST
+from saml2.client import Saml2Client
+from saml2.validate import ResponseLifetimeExceed
+
 BASIC_AUTH_WEBUI_REALM = "tracim"
 TRACIM_API_KEY_HEADER = "Tracim-Api-Key"
 TRACIM_API_USER_LOGIN_HEADER = "Tracim-Api-Login"
 AUTH_TOKEN_QUERY_PARAMETER = "access_token"
 CLIENT_TOKEN_HEADER = "X-Tracim-ClientToken"
 
 
+class SAMLSecurityPolicy:
+    def __init__(self, app_config: CFG, configurator: Configurator) -> None:
+        self.app_config = app_config
+        self._session_helper = SessionAuthenticationHelper()
+        # TODO - SGD - 2023-07-26 - Add views for sso/slo/acs/sls
+        #  The routes could also be defined through a controller (see Controller class)
+        #  even if it added-value is small.
+
+        # TODO - M.L - 2023/09/05 - Change this to proper config loading
+        self._load_settings(configurator)
+        _config = configurator.get_settings().get('pyramid_saml')
+        with open(_config.get('saml_path'), 'r') as config_file:
+            config_data = json.load(config_file)
+            self.saml_config = Saml2Config()
+            self.saml_config.load(config_data)
+            attribute_map = {
+                "EmailAddress": "user_email",
+                "UserID": "user_id",
+                "FirstName": "first_name",
+                "LastName": "last_name",
+            }
+            self.saml_config.allow_unknown_attributes = True
+            self.saml_config.attribute_map = attribute_map
+        self._metadata_response = Response(content_type='application/xml')
+        self._metadata_response.text = create_metadata_string(config=self.saml_config,
+                                                              configfile=None).decode()
+        self.saml_client = Saml2Client(config=self.saml_config)
+
+        for _, data in self.saml_client.config.vorg.items():
+            if data.common_identifier not in app_config.IDP_LIST:
+                app_config.IDP_LIST.append(data.common_identifier)
```

Comment:
This is the process of adding the configuration from `backend/settings_saml2.json` (as well as adding mandatory parameters, such as `self.saml_config.allow_unknown_attributes = True`

`self.saml_config.attribute_map = attribute_map` and the definition of `attribute_map` might be unnecessary

Some preparation for the metadata endpoint is also done, it's done to compute the metadata response (and fetch metadata from the remote IdPs) once at startup, to prevent slowdowns of the request handling

The list of IdPs is also added to the `app_config` under the `IDP_LIST` attribute